File: problems/74-search-a-2d-matrix.scala
// Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 3
// Output: true

// Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 13
// Output: false

//> using file helpers.scala

package p74

import leetcode.Helpers.pp
import scala.annotation.tailrec

// object Solution {
//   def searchMatrix(matrix: Array[Array[Int]], target: Int): Boolean = {
//     val rows = matrix.length
//     val cols = matrix(0).length
//     var left = 0
//     var right = rows * cols - 1
//     while (left <= right) {
//       val mid = left + (right - left) / 2
//       val value = matrix(mid / cols)(mid % cols)
//       if (value == target) return true
//       else if (value < target) left = mid + 1
//       else right = mid - 1
//     }
//     false
//   }
// }

object Solution {
  def searchMatrix(matrix: Array[Array[Int]], target: Int): Boolean = {
    val rows = matrix.length
    val cols = matrix(0).length
    val left = 0
    val right = rows * cols - 1

    @scala.annotation.tailrec
    def go(left: Int, right: Int): Boolean = {
      if (left > right) return false
      val mid = left + (right - left) / 2
      val value = matrix(mid / cols)(mid % cols)
      if (value == target) return true
      else if (value < target) go(mid + 1, right)
      else go(left, mid - 1)
    }

    go(left, right)
  }
}

object Main {
  def main(args: Array[String]): Unit = {
    val matrix1 =
      Array(Array(1, 3, 5, 7), Array(10, 11, 16, 20), Array(23, 30, 34, 60))
    val target1 = 3
    println("Example 1")
    println(s"Input: matrix=${pp(matrix1)}, target=${pp(target1)}")
    println(s"Output: ${pp(Solution.searchMatrix(matrix1, target1))}")
    println(s"Correct: ${pp(true)}")

    val matrix2 =
      Array(Array(1, 3, 5, 7), Array(10, 11, 16, 20), Array(23, 30, 34, 60))
    val target2 = 13
    println("Example 2")
    println(s"Input: matrix=${pp(matrix2)}, target=${pp(target2)}")
    println(s"Output: ${pp(Solution.searchMatrix(matrix2, target2))}")
    println(s"Correct: ${pp(false)}")
  }
}

File: problems/74-search-a-2d-matrix.py
# Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 3
# Output: true

# Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 13
# Output: false

from __future__ import annotations
from typing import List


def search(self, nums: List[int], target: int) -> int:
    left = 0
    right = len(nums) - 1

    while left <= right:
        middle = left + ((right - left) // 2)
        if nums[middle] < target:
            left = middle + 1
        elif nums[middle] > target:
            right = middle - 1
        else:
            return middle
    return -1


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows = len(matrix)
        cols = len(matrix[0])

        def get_row_and_col(index):
            col = index % cols
            row = index // cols
            return (row, col)

        left = 0
        right = rows * cols - 1

        while left <= right:
            middle = left + ((right - left) // 2)
            row, col = get_row_and_col(middle)
            curr = matrix[row][col]
            if curr < target:
                left = middle + 1
            elif curr > target:
                right = middle - 1
            else:
                return True

        return False


solution = Solution()

input_1 = {"matrix": [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]], "target": 3}
print("Example 1")
print("Input:", input_1)
print("Output:", solution.searchMatrix(**input_1))
print("Correct:", True)

input_2 = {"matrix": [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]], "target": 13}
print("Example 2")
print("Input:", input_2)
print("Output:", solution.searchMatrix(**input_2))
print("Correct:", False)
